### PR TITLE
Token distribution

### PR DESCRIPTION
**Pull request by avclarke**

### CODE REVIEW

File: migrations/2_deploy_contracts.js [hunk 1]
```diff
@@ -2,12 +2,15 @@
...
 const MainframeStake = artifacts.require('MainframeStake')
+const MainframeTokenDistribution = artifacts.require('MainframeTokenDistribution')
 const MainframeEscrow = artifacts.require('MainframeEscrow')
 
 module.exports = (deployer, network) => {
   deployer.deploy(MainframeToken).then(() => {
     return deployer.deploy(MainframeEscrow, MainframeToken.address)
   }).then(() => {
     return deployer.deploy(MainframeStake, MainframeEscrow.address)
+  }).then(() => {
```
Comment (PaulLeCam) [inline migrations/2_deploy_contracts.js]: I think this deployment could be done in parallel to the others as it has a dependency on `MainframeToken` but not the other contracts, ex:

```js
const deployEscrowAndStake = () => {
  return deployer.deploy(MainframeEscrow, MainframeToken.address).then(() => {
    return deployer.deploy(MainframeStake, MainframeEscrow.address)
  })
}

deployer.deploy(MainframeToken).then(() => {
  return Promise.all([
    deployEscrowAndStake(),
    deployer.deploy(MainframeTokenDistribution, MainframeToken.address),
  ])
})
```
This way the deployment of `MainframeTokenDistribution` only waits for the deployment of `MainframeToken`, not the other 2.

File: test/TestMainframeTokenDistribution.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+const MainframeDistribution = artifacts.require('./MainframeTokenDistribution.sol')
+const MainframeToken = artifacts.require('./MainframeToken.sol')
+const utils = require('./utils.js')
+
+contract('MainframeAirDrop', (accounts) => {
```
Comment (mosic) [inline test/TestMainframeTokenDistribution.js]: `s/MainframeAirDrop/MainframeTokenDistribution/`

File: contracts/MainframeTokenDistribution.sol [hunk 3]
```diff
@@ -0,0 +1,37 @@
+pragma solidity ^0.4.18;
+
+import "zeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./MainframeToken.sol";
```
Comment (howleysv) [inline contracts/MainframeTokenDistribution.sol]: You could reference `ERC20` here instead to indicate that it doesn't rely on any special functionality of the mainframe token contract, similar to the escrow contract.

File: contracts/MainframeTokenDistribution.sol [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  function distributeTokens(address tokenOwner, address[] recipients, uint[] values) onlyOwner external {
+    require(validate(tokenOwner, recipients, values));
+    for(uint i = 0; i < recipients.length; i++) {
+      if(values[i] > 0) {
+        require(token.transferFrom(tokenOwner, recipients[i], values[i]));
+        TokensDistributed(recipients[i], values[i]);
+        totalDistributed += values[i];
+      }
+    }
+  }
```
Comment (howleysv) [inline contracts/MainframeTokenDistribution.sol]: We may as well add the token drain too since we have it in all of our other contracts

**Pull request closed without merging**